src/templates/head-meta.js

```diff
@@ -127,10 +127,37 @@ export function nuxtI18nHead ({ addDirAttribute = false, addSeoAttributes = fals
     const canonicalPath = currentRoute ? currentRoute.path : null
 
     if (canonicalPath) {
+      let href = toAbsoluteUrl(canonicalPath, baseUrl)
+
+      if (currentRoute && Array.isArray(canonicalQueries) && canonicalQueries.length) {
+        const currentRouteQueryParams = currentRoute.query
+        const params = new URLSearchParams()
+        for (const queryParamName of canonicalQueries) {
+          if (queryParamName in currentRouteQueryParams) {
+            const queryParamValue = currentRouteQueryParams[queryParamName]
+
+            if (Array.isArray(queryParamValue)) {
+              queryParamValue.forEach(v => params.append(queryParamName, v || ''))
+            } else {
+              params.append(queryParamName, queryParamValue)
+            }
+          }
+        }
+        let queryString = params.toString()
+
+        if (queryString) {
+          // URLSearchParams.append won't let us append just a key, so remove null and undefined values
+          queryString = queryString.replace('=undefined', '')
+          queryString = queryString.replace('=null', '')
```

Comment:
I have said to remove the string type check because according to VueRouter types, the queryParamValue can only be a string or an array of string or null. But if that's not the case and the types are wrong then I'm fine with instead changing the `params.append(queryParamName, queryParamValue)` line to `params.append(queryParamName, queryParamValue || '')` for example.

Reply:
@rwwagner90 you could have just used suggested logic... not sure why you didn't use it at first instead changed it multiple times and are now back to the initial one.

Reply:
@divine I have explained my reasons, and I do not appreciate the hostility here. Again, the suggestions of having a plain else block does not work. A url of `?foo` will end up being `?foo=null`. Your code did not handle this case either.

Reply:
> I have said to remove the string type check because according to VueRouter types, the queryParamValue can only be a string or an array of string or null. But if that's not the case and the types are wrong then I'm fine with instead changing the `params.append(queryParamName, queryParamValue)` line to `params.append(queryParamName, queryParamValue || '')` for example.

@rchl Thank you for clarifying! I believe the issue is when it's undefined like `?foo` as you say here undefined is not a possible value, so it converts it to null, then we end up with `?foo=null` because URLSearchParams does not support setting keys with no values.

Reply:
@rchl no unfortunately that does not help. `params.append(queryParamName, queryParamValue || '')` just means you will end up with `nuxt-app.localhost/?foo=` so you have an extra equals there and no value and you requested that when the value is `undefined` we instead have `?foo` not `?foo=`. Personally, I wouldn't support the `?foo` case where there is a key and no value, but I added the hacks to support this:

```js
       if (queryString) {
          // URLSearchParams.append won't let us append just a key, so remove null and undefined values
          queryString = queryString.replace('=undefined', '')
          queryString = queryString.replace('=null', '')

          href = `${href}?${queryString}`
        }
```

Reply:
I'm fine with including the `=` because it won't make any difference from the perspective of parsing the query params.

Reply:
I've pushed a fix for the types.